File: typescript/examples/tools/convert-to-file/convert-to-file-decorator.ts
import {
  MatimoInstance,
  setGlobalMatimoInstance,
  tool,
  getGlobalApprovalHandler,
  type ApprovalRequest,
} from '@matimo/core';

/**
 * Create an interactive approval callback for file conversion operations.
 */
function createApprovalCallback() {
  return async (request: ApprovalRequest): Promise<boolean> => {
    const isInteractive = process.stdin.isTTY;

    console.info('\n' + '='.repeat(70));
    console.info('🔒 APPROVAL REQUIRED FOR FILE CONVERSION');
    console.info('='.repeat(70));
    console.info(`\n📋 Tool: ${request.toolName}`);
    console.info(`   ${request.params.source_format} -> ${request.params.target_format}`);

    if (!isInteractive) {
      console.info('\n❌ REJECTED - Non-interactive environment (no terminal)');
      console.info('\n💡 To enable auto-approval in CI/scripts:');
      console.info('   export MATIMO_AUTO_APPROVE=true');
      console.info('\n' + '='.repeat(70) + '\n');
      return false;
    }

    // Non-interactive example environments should set MATIMO_AUTO_APPROVE=true.
    console.info('   ✅ Auto-approved (see MATIMO_AUTO_APPROVE)');
    console.info('='.repeat(70) + '\n');
    return true;
  };
}

/**
 * Example: convert_to_file tool using the @tool decorator pattern.
 */
class FileConverter {
  @tool('convert_to_file')
  async convert(content: string, source_format: string, target_format: string): Promise<unknown> {
    // Decorator automatically intercepts and executes via Matimo
    return undefined;
  }
}

async function decoratorExample() {
  const matimo = await MatimoInstance.init({ autoDiscover: true });
  setGlobalMatimoInstance(matimo);

  const approvalHandler = getGlobalApprovalHandler();
  approvalHandler.setApprovalCallback(createApprovalCallback());

  console.info('\n' + '='.repeat(70));
  console.info('🚀 Convert To File Tool - Decorator Pattern Example');
  console.info('='.repeat(70));

  const converter = new FileConverter();

  try {
    console.info('\n1️⃣  CONVERTING CSV TO JSON');
    console.info('-'.repeat(70));
    const csv = 'name,role\nAda,Mathematician\nAlan,Computer Scientist\n';
    const result = await converter.convert(csv, 'csv', 'json');
    if (result) {
      const decoded = Buffer.from((result as any).file_base64, 'base64').toString('utf8');
      console.info('✅ MIME type:', (result as any).mime_type);
      console.info('📝 JSON output:', decoded);
    }
    console.info('---\n');

    console.info('✅ Decorator example completed successfully');
  } catch (error: any) {
    console.error('❌ Error:', error.message);
  }
}

decoratorExample();

File: python/examples/native/convert_to_file/convert_to_file_decorator.py
#!/usr/bin/env python3
"""
============================================================================
CONVERT_TO_FILE TOOL — DECORATOR PATTERN
============================================================================

PATTERN: @tool Decorator
────────────────────────────────────────────────────────────────────────────
Wraps Matimo tool calls in a class using the @tool("tool-name") decorator.
The decorator intercepts method calls and routes them through Matimo
automatically — the method body is never executed.

Use this pattern when:
  ✅ Building class-based applications or services
  ✅ Encapsulating content-conversion logic in strongly-typed wrappers
  ✅ Combining multiple tools in a single service layer
  ✅ Object-oriented design

SETUP:
────────────────────────────────────────────────────────────────────────────
  No API key required — convert_to_file is a built-in core tool.

USAGE:
────────────────────────────────────────────────────────────────────────────
  export MATIMO_AUTO_APPROVE=true
  uv run python native/convert_to_file/convert_to_file_decorator.py

============================================================================
"""

import asyncio
import base64
from pathlib import Path

from dotenv import load_dotenv

from matimo import Matimo
from matimo.decorators import set_global_matimo_instance, tool

load_dotenv(Path(__file__).parent.parent.parent / ".env")


# ───────────────────────────────────────────────────────────────────────────
# Service class — each method is auto-routed to the matching Matimo tool
# ───────────────────────────────────────────────────────────────────────────

class FileConverter:
    """High-level content-conversion service using the @tool decorator pattern."""

    @tool("convert_to_file")
    async def convert(self, content: str, source_format: str, target_format: str) -> dict:
        """
        Decorator auto-calls matimo.execute('convert_to_file', {...}).

        Args:
            content: The source content to convert
            source_format: 'json' | 'csv' | 'markdown' | 'text'
            target_format: 'pdf' | 'docx' | 'csv' | 'json' | 'txt'

        Returns:
            output_path / file_base64 / mime_type / size_bytes
        """
        ...


# ───────────────────────────────────────────────────────────────────────────
# Main
# ───────────────────────────────────────────────────────────────────────────

async def main() -> None:
    print("\n╔════════════════════════════════════════════════════════╗")
    print("║     Convert To File Tool — Decorator Pattern            ║")
    print("║     (@tool decorators for automatic execution)          ║")
    print("╚════════════════════════════════════════════════════════╝\n")

    print("🚀  Initializing Matimo…")
    matimo = await Matimo.init(auto_discover=True)
    set_global_matimo_instance(matimo)
    print("✅  Matimo initialized\n")

    converter = FileConverter()

    try:
        print("1. Converting CSV to JSON\n")
        csv_content = "name,role\nAda,Mathematician\nAlan,Computer Scientist\n"
        result = await converter.convert(csv_content, "csv", "json")
        if result.get("success"):
            decoded = base64.b64decode(result["file_base64"]).decode("utf-8")
            print(f"✅  MIME type: {result.get('mime_type')}")
            print(f"📝  JSON output:\n{decoded}")
        else:
            print(f"Error: {result.get('error')}")
        print("---\n")

        print("✅  Decorator example completed successfully")
    except Exception as error:
        print(f"❌  Error: {error}\n")


if __name__ == "__main__":
    asyncio.run(main())
